### ISSUE TITLE
[hmac] V2(S) Signoff

### ISSUE THREAD

**Issue by andreaskurth** [lowRISC/opentitan#22471]

**Comment by martin-velay** [comment]
### Commits since HMAC D2(S) Signoff (https://github.com/lowRISC/opentitan/issues/20996)
---
`git rev-parse HEAD`

https://github.com/lowRISC/opentitan/commit/ea5bce9ed692af726da4729039c4cbe1fba64ee7

`git log d7483116b80bf56ce2b8e665c679872de3f3794d..HEAD --oneline hw/ip/hmac`

99ce062a64 [hmac, dv] Fix `ValidHmacEnConditionAssert`
01a208901a [hmac,rtl/dv] DV synchronization and error handling fixes
f489ff703d [hmac,rtl] HMAC FSM RTL fixes to handle SAR
7d8ace24cb [dv] Remove default value for wait_to_issue_reset argument
6c51bc0da2 [hmac,dv] DV fixes for block-level tests
ee5f4a1abf [sram_ctrl] Add readback feature
f9f6fe5d06 [hmac,csrs/rtl] Add key_swap field and expose hmac idle status
aace3c0617 [hmac,rtl/dv] RTL and DV fixes for digest loading and error handling
76b1c7498c [hmac,dv] DV synchronization fixes for do_read_check
c1fa3221d0 [hmac,rtl] Minor HMAC fixes for secret key wiping
:...skipping...
99ce062a64 [hmac, dv] Fix `ValidHmacEnConditionAssert`
01a208901a [hmac,rtl/dv] DV synchronization and error handling fixes
f489ff703d [hmac,rtl] HMAC FSM RTL fixes to handle SAR
7d8ace24cb [dv] Remove default value for wait_to_issue_reset argument
6c51bc0da2 [hmac,dv] DV fixes for block-level tests
ee5f4a1abf [sram_ctrl] Add readback feature
f9f6fe5d06 [hmac,csrs/rtl] Add key_swap field and expose hmac idle status
aace3c0617 [hmac,rtl/dv] RTL and DV fixes for digest loading and error handling
76b1c7498c [hmac,dv] DV synchronization fixes for do_read_check
c1fa3221d0 [hmac,rtl] Minor HMAC fixes for secret key wiping
5a74a15fb1 [hmac/rtl] Fix idle_o when told to stop
e71da1ced8 [hmac/rtl] Wait for digest of complete block when stopping
dbe6169068 [prim_sha2_pad,rtl] Go to idle (without padding) when told to stop
e01609eb0d [hmac,doc] Documentation fixes
128d609a2a [hmac,dv] Scoreboard checks for different digest sizes
e120a95bb6 [hmac/prim_2,rtl] Do not clear redundant digest values
ed04dc9cf4 [hmac, doc] Update HMAC context saving guide
f189ae8354 [hmac, doc] Update programmers guide for streaming
5bc6e0a96f [hmac, rtl] Fix indexing bug/typo
ab878b5d35 [hmac/dv] Cleanups and coverage fixes/improvements
e2eff20902 [hmac,dv] Extend DV to test invalid configurations
c859a39662 [hmac/prim_sha2,rtl] Implement SW error for invalid HMAC config
4bc15708cd [prim_sha2,rtl/dv] Fix secret value wiping
0d0ce1ad80 [hmac/dv] Cleanups without functional impact
dc848b3e2a [sram_ctrl,prim,rtl] Fix tlul_we -> ready timing path.
8295903d7e [bazel,autogen_hjson] Split C and rust header generation
30d7e787c7 (tag: Earlgrey-PROD-M2-RC5, tag: Earlgrey-PROD-M2) Add the project name to the copyright header
d01864e18e Fix or waive Python lint errors and warnings
f4c2bb9030 Remove trailing whitespaces

### Issues closed since HMAC D2(S) Signoff (https://github.com/lowRISC/opentitan/issues/20996)
---
[Issues closed since HMAC D2(S) Signoff](https://github.com/lowRISC/opentitan/issues?q=is%3Aissue+is%3Aclosed+closed%3A%3E2024-03-28+hmac+)

**Comment by moidx** [comment]
Discussed moving as P1 to M5 due to RTL risk analysis performed during last issue triage.

**Comment by vogelpi** [comment]
Alright, AFAIK all outstanding PRs have been merged and issues addresses. @martin-velay is currently investigating an issue with FSM coverage but after that we should be able to close V2S.

**Comment by martin-velay** [comment]
Here are the latest updates: https://github.com/lowRISC/opentitan/issues/23683

Coverage results in this comment: https://github.com/lowRISC/opentitan/issues/23683#issuecomment-2210190963. This includes the UNR exclusions from this PR https://github.com/lowRISC/opentitan/pull/23916 which is not merged yet.

The FSM cov is still a bit lower than 90%, and here is the reason: https://github.com/lowRISC/opentitan/issues/23683#issuecomment-2208903846

**Comment by vogelpi** [comment]
Thanks @martin-velay , I've feed back on the FSM coverage issue. I believe this should be coverable but it may need very precise timing.

I've merged the updated UNR file.

**Comment by andreaskurth** [comment]
Thanks for the V2S review, @martin-velay, and all your work on this, also to @gdessouky!

I reviewed the following commits, which change the RTL of HMAC and/or the SHA2 prims it instantiates:
* 5084b38250 [hmac] Coding style and minor fixes
  --> lint and coding style fixes
* 30d7e787c7 Add the project name to the copyright header
  --> no functional change
* dc848b3e2a [sram_ctrl,prim,rtl] Fix tlul_we -> ready timing path.
  --> no functional change to HMAC or SHA2
* 4bc15708cd [prim_sha2,rtl/dv] Fix secret value wiping
  --> fix for issue #22555
* c859a39662 [hmac/prim_sha2,rtl] Implement SW error for invalid HMAC config
  --> fix for issue #22312
* 5bc6e0a96f [hmac, rtl] Fix indexing bug/typo
  --> fix for issue #23014
* e120a95bb6 [hmac/prim_2,rtl] Do not clear redundant digest values
  --> fix wiping with random value for upper digest registers in SHA2_256 mode
* a7cb9a484e [prim_sha2_pad,rtl] Fix setting of digest mode when continuing
  --> fix for save & restore feature
* 6e6c14b56d [prim_sha2_pad,rtl] Refactor comparison on tx_count and msg len into signal
  --> small refactoring to enable subsequent commit
* dbe6169068 [prim_sha2_pad,rtl] Go to idle (without padding) when told to stop
  8fad05fba7 [prim_sha2_pad,rtl] Signal msg feed complete also when stopping
  e71da1ced8 [hmac/rtl] Wait for digest of complete block when stopping
  --> fixes for save & restore feature
* 5a74a15fb1 [hmac/rtl] Fix idle_o when told to stop
  --> fix for issue #23036
* c1fa3221d0 [hmac,rtl] Minor HMAC fixes for secret key wiping
  --> fix for secret wiping
* aace3c0617 [hmac,rtl/dv] RTL and DV fixes for digest loading and error handling
  --> fix digest loading and improve error handling
* f9f6fe5d06 [hmac,csrs/rtl] Add key_swap field and expose hmac idle status
  --> resolve issue #23158
* ee5f4a1abf [sram_ctrl] Add readback feature
  --> no functional change to HMAC or SHA2
* 9e6feec006 [kmac,prim_sha2/rtl] Remove unused `conv_endian64` function
  --> code removal for lint cleaning
* 0742a8e0ad [kmac,prim_sha2/rtl] Fix lint error in `conv_endian32`
  --> lint cleanup, no functional change
* f489ff703d [hmac,rtl] HMAC FSM RTL fixes to handle SAR
  --> fix for save & restore feature
* 01a208901a [hmac,rtl/dv] DV synchronization and error handling fixes
  --> fix loading from message FIFO
* 99ce062a64 [hmac, dv] Fix `ValidHmacEnConditionAssert`
  --> fix issue #23348
* 32d52b8d41 [hmac,rtl] Fix Save & Restore FSM bug
  --> fix for save & restore feature
* 37bcf1ef9f [hmac/lint] Remove unused localparam
  --> lint cleanup, no functional change
* d79a3bb4f8 [hmac,rtl] Coding cosmetics fixes
  --> no functional change
* 85baec433f [hmac,dv] Minor DV assertion fix
  --> no functional change
* d851a38dc7 [hmac] Avoid inference of latch
  --> prevent unintended latch

Those RTL changes LGTM.

Test pass rates are excellent overall (today 100% including V3 tests!), though the `fifo_empty_status_interrupt` testpoint currently has no tests assigned with it. Since PR #23749, the scoreboard includes checks for this feature, so we just need to list the appropriate tests.

Coverage numbers are very good (94 to 100%) except FSM, which is known and for which an exclusion is being added.

None of the open issues prevents us from signing HMAC off at V2S in my view.

In conclusion, once the exclusions for FSM coverage are in and the `fifo_empty_status_interrupt` testpoint has tests assigned with it, I agree with signing HMAC off at V2S.

**Comment by gdessouky** [comment]
Exclusions for the FSM coverage per our discussion here https://github.com/lowRISC/opentitan/issues/23683 have been added in https://github.com/lowRISC/opentitan/pull/23948.

**Comment by gdessouky** [comment]
FSM coverage is now at 97% with the additional FSM exclusions.
